### ISSUE TITLE
Switch to backend-generated UI

### ISSUE THREAD

**Issue by umputun** [umputun/remark42#825]
Just an idea I'd like to explore and will apricate feedbacks from both backend and frontend devs.

Currently, we have a decent separation between UI and the backend with REST-like API. This is all fine and standard; however, in multiple places the separation bites us, and we are forced to do odd things, like infamous sed preparing HTML. Another issue is two separate places to configure things. And the last problem - we have a constant backlog of UI tasks, and the situation doesn't get any better. One more thing - with an endless number of various npm modules involved in the build/deploy/ui-functionality we are constantly racing against security issues and have to update things way too often just to address new vulnerabilities.

From another hand, our UI is not that advanced — all we need to do just 3 "widgets" with a minimal level of interactivity. Counter widget is trivial, last comments kind of static thingy and the primary comments tree already delegates some complicated parts (like preview generation) to server. The ideal solution could be fully static HTML generated by the server. This is very tempting and, theoretically, can be extremely fast, CDN friendly, and all other goodies we get from the static content. The issue with this solution - it may require a significant amount of extra generation, i.e. we will need all HTML for all sorts, all pages, and different pages/views for the different access levels. It could also be challenging from auth/security POV.

I think entirely static rendering will be hard to do right and may limit us significantly. It is why something in between, namely [htmx](https://htmx.org). We could render UI components dynamically on the server-side with this library without the need for the second language (js/ts) and a dedicated UI framework. Well, htmx is a kind of js framework but sort of а different beast. The idea of such a "dynamic server-side rendering" is something unusual these days, but not necessarily a bad thing. I'm not sure how all of this can be integrated, and if we could use it as-is or write some extensions, but generally speaking to me, this direction looks very promising.

cc: @akellbl4 @paskal

**Comment by umputun** [comment]
I have very limited practical experience with htmx. All I did was a small project with a relatively simple, but dynamic UI for a monitoring/orchestration system. My impression after this project was mostly positive. It can be used and allowed me to make the frontend (bootstrap based) without much headache. 

Some problems I run to:
- Initially, it feels very unusual to write component controllers returning HTML snippets and rendering from templates
- The right structure still somewhat unclear, I mean how to organize things, where to put controllers, templates, and so on
- Sometimes it is not easy to pick the right scope for the component 
- Stateful parts of UI need to keep the state on the server-side somehow

**Comment by paskal** [comment]
As we won't get rid of our API anyway, I think it's worth trying to implement that on a separate endpoint as an experiment.

I have a lot of concerns which can't be answered before we have a working prototype, I don't trust pure model as htmx describe it on the main page: I saw that they themselves have [an example with input verification](https://htmx.org/examples/click-to-edit/) which is done on a client:
![image](https://user-images.githubusercontent.com/712534/100559956-9af23a00-32b4-11eb-9855-6d219596e042.png)
Using htmx plus plain JS for clearly client stuff could be the sane alternative to the complexity of the UI we have at the moment.

**Comment by akellbl4** [comment]
I always support keeping things as simple as possible but our comments logic really complex and huge (rendering comments tree, updating and managing comments, etc). Widgets with last comments and counter are small and simple.

What we'll be diffrent:
- We will lose testing environment
- immature technology
- htmlx size is 9kb against preact 3kb, but maybe our final code would be smaller
- htmlx uses custom HTML attributes (minor thing but I don't like such an approach)
What we'll keep:
- almost the same number of production dependencies (we could reduce them already, just need to pay some time for it)
- we will keep some build stage for frontend as well just because we will need to bundle our code from modules to a single file (or creating chunks) and transpile it and obfuscate it for the smallest size

Other thoughts about htmlx: I haven't seen any project on it, their examples really simple as well (they could be implemented with pure js in 20 lines)

Arguments above just the first cons that came to me from the first look.

> "dynamic server-side rendering" is something unusual these days
I agree server-side rendering is a great thing and almost all of the projects what I know use JS for it.
I use SSR every day on my work project. It works on NextJs and renders my React (in our case Preact) application on server.

**Comment by akellbl4** [comment]
I have a few ideas about how or frontend could be simplier and smaller.
We could rid of around half of the production js dependencies and more than that from dev dependencies.
Also we could rid of `sed` with "shallow server rendering" I mean, we could send HTML via Go server and put all initialization info such as initial configuration and comments data (in JSON) and then render it without delay on a client.

**Comment by umputun** [comment]
Good to hear the other side ;) A few questions:

> We will lose testing environment

What does it mean? Testing of what exactly? The code for the actual components will be the regular go code which we know how to test

> almost the same number of production dependencies

Not sure I get it. What dependencies did you mean? Let's say we can achieve the goal and have no-js (or almost no js) code involved and everything dynamic (as well as static) will be handled by the server. Unless I'm missing something big here, I don't see why we will need any dependencies 

> we will keep some build stage for frontend as well just because we will need to bundle our code from modules to a single file (or creating chunks) and transpile it and obfuscate it for the smallest size

 I don't understand this one either. What exactly will we need to build? If all components made as controllers + templates the only build-related part (for binary deployment) will be packing templates with statik or smth similar.

Just to clarify: I'm not considering htmx approach as some kind of node.js replacement rendering UI, but rather as a helper to provide the dynamic functionality we have/want. This is very close to the good-old JSP but with a little bit of ajax (not even sure if this term in use anymore). Maybe this goal is too ambitious, not sure. But what makes me optimistic - even now server-side doing a lot of things related to the presentation, for example, it renders the tree of comments. It is not exactly in the form we can use in the browser but rendering to divs not such a big difference from the hierarchal json we currently serve.

To me, the most challenging part is how to deal with state management. I think such a thin client may have some issues to keep the state and communicate it back, but probably this can be resolved.

**Comment by umputun** [comment]
> htmlx uses custom HTML attributes (minor thing but I don't like such an approach)

I've read somewhere about the use of data (or something similar) attributes instead of the custom set. Not sure if this is related to what you hade in mind, but in that discussion peps considered it as a more standard way

**Comment by umputun** [comment]
> Another huge con is the support of internationalization.

what is the issue with internationalization? I can think of multiple ways to achieve it on the server-side, quick googling gave me this https://phrase.com/blog/posts/internationalization-i18n-go/ for example.

**Comment by akellbl4** [comment]
> What does it mean? Testing of what exactly? The code for the actual components will be the regular go code which we know how to test

Now we have tests for frontend components and we test their behavior, some sort of unit tests for UI. But I think we could switch to E2E tests they are more complicated to write but more reliable.

> What dependencies did you mean?

Okay, I haven't understood htmlx fully from the first look. It's a bit different than I thought. Looks like we will need only this lib and all our logic will migrate to HTML with attributes what could be quite difficult to support. I think we could make a small example to find out and one of the complex things in my opinion it is the login form:
- Dynamically hiding and showing icons
- Dropdowns
- Steps (Credentials step and then token step)

> I've read somewhere about the use of data (or something similar) attributes instead of the custom set. Not sure if this is related to what you hade in mind, but in that discussion, peps considered it as a more standard way

Yep, `data-*` is a good way to put some sort of custom attributes to DOM.

> what is the issue with internationalization?

If all interface always renders on the server it isn't the issue.


The summary of this comment:
Rendering everything on the server it's could be good and bad at the same time.
First of all, because we need to request every part of the interface from the server or we need to write some JS to show/hide or change something in the interface on the frontend (again maybe I'm wrong and because I read docs of htmlx shallowly).
The messiest thing that I found is validation (js right in HTML attributes) https://htmx.org/docs/#validation
At the same time, SSE looks interesting but I very curious how it could work for a huge amount of comments.
Last but not least I can see how their approach smashes the wall because they have an extension for rendering templates from JSON server responses (https://htmx.org/extensions/client-side-templates/) maybe it was created for some sort of compatibility and I'm wrong.
htmlx also works differently from modern SSR approaches because now it works like render the first request as HTML with JS templates and then request JSON and render it with the same code on the client-side.
In the end, I want to say If we use htmlx it will be "no js" just from a code point of view :) because no-js means no js at all, but we will use JS lib which operates by HTML attributes. And yep I could support this approach just because I'm interested in writing some Go code instead of JS/TS :)

P.S. I don't have strong points for denying this approach but also I don't have them to support it. I could see htmlx is an interesting but a really controversial approatch.

**Comment by paskal** [comment]
Revisiting this after a week working inside the frontend build. Some measurements, and an attempt at the objections raised above that were never really resolved.

### What the widget actually is today

157 non-test source files, 8,745 lines, 29 CSS module files, and a Redux store of 1,052 lines. Fifteen runtime dependencies against **83 devDependencies**.

That last ratio is the finding. The interface is not complicated; the machinery around it is.

### The security-churn point holds up

The original post noted we are "constantly racing against security issues". That is measurable now: `pnpm audit` reported 23 vulnerabilities in the frontend this week, and every one was build or test tooling. None shipped in the widget bundle. Clearing them meant raising override floors on nine packages. That work is a recurring tax paid purely for machinery.

### A correction to how this gets framed

I want to push back on one assumption, including in my own earlier thinking: Preact is not the expensive part.

Preact is imported directly in around 50 non-test files, but almost entirely for `h` and JSX types. Those files do not care what renders them. What costs us is that **Preact has to impersonate React**: `react` and `react-dom` are npm aliases to a compat shim, and that alias exists for exactly two consumers, `react-redux` (13 files) and `react-intl` (32). No file in the app imports `react` or `react-dom` directly. `react-redux` also drags a real `react` resolution into the lockfile transitively, through `use-sync-external-store`.

Every hard problem in the recent preact 10.6 to 10.29 upgrade came from that impersonation:

- `preact/compat` silently remaps `onFocus` to `focusin` and `onBlur` to `focusout`, so a test dispatching a plain `focus` event never reaches the handler, with no error
- `@testing-library/preact` rewrites `change` to `input` for every element once it sees a compat vnode, which is right for `<input>` and wrong for `<select>`, and broke a test unrelated to the change
- the third-party `@preact/compat` package sat at 17.x while preact was on 10.29
- all 78 TypeScript errors in that upgrade clustered on the alias

Dropping `react-redux` and `react-intl` removes the alias, the `paths` entries in `tsconfig.json`, the compat dependency and that entire class of bug, without touching the rendering model at all.

### On the objections from earlier in the thread

**"We will lose testing environment."** Partly true, partly already lost. We currently run three test libraries at once: `enzyme` via the abandoned `enzyme-adapter-preact-pure` in 3 files under `app` plus `jest.setup.ts`, `@testing-library/preact` in 16, and `@testing-library/preact-hooks` in 1. Server-rendered components become ordinary Go handler tests, which is a better story than maintaining those three.

**"Internationalization support."** The strongest objection, and it survives. It is bounded though: five API shapes in use (`FormattedMessage` 53 occurrences, `useIntl` 43, `defineMessages` 30, `IntlShape` 20, `IntlProvider` 20), catalogues already extracted to JSON, extraction tooling already ours in `tasks/`. Message lookup maps onto Go templates straightforwardly; plurals and date formatting are the parts needing real design rather than mechanical translation.

**"htmx is 9 kB against preact 3 kB."** True, and not the relevant comparison. It is htmx against preact plus compat plus react-redux plus react-intl and everything they pull in. `remark.mjs` is currently 78 kB gzipped.

**"We will keep a build stage anyway."** Much less of one. The webpack build currently runs source through `ts-loader` and then `babel-loader`, while jest uses `@swc/jest`, so three transform paths are configured across two pipelines. Server-rendered templates need none of them.

### What genuinely cannot move to the server

Worth naming precisely so nobody plans around them:

1. **The embed script.** `embed.mjs`, 2.2 kB, creates the iframe and syncs height, theme and anchor scrolling over postMessage. It stays hand-written JS.
2. **Auth.** Eleven default OAuth providers, arbitrary custom provider strings, plus email and anonymous form flows. OAuth popup windows that postMessage back to the opener are not something htmx addresses; form posts and redirects are.
3. **The composer.** Markdown preview could round-trip to the server naturally. Image paste and drag upload with progress, textarea autosize, mention autocomplete and draft persistence stay client-side.

So the end state is not "no JS". It is a few hundred lines for embed, auth and composer, against 8,745 lines today.

### The head start is bigger than the original post suggested

`backend/app/store/formatter.go` already renders markdown with blackfriday and chroma, and `comment.go` sanitises the result with bluemonday's UGC policy, keeping the raw markdown in `Orig`. Comment bodies already reach clients as sanitised HTML. Server-rendering the thread means templating structure around content the backend produces today, not building a rendering pipeline.

`frontend/packages/api` is also already a standalone typed client with zero runtime dependencies and no preact or react imports, so anyone wanting a Svelte or Vue frontend is unblocked regardless of what we decide here.

### Suggested sequencing

I would not start with htmx. By value per unit of work:

1. **Remove the React alias.** The `react-redux` coupling is concentrated in `app/store/index.ts`, which exports the typed hooks, plus the call sites. No change to the rendering model.
2. **Collapse the toolchain**, webpack plus babel plus swc plus jest onto one bundler and one test runner. `packages/api` already uses vitest, so there is in-repo precedent. This is where most of the 83 devDependencies live.
3. **Stabilise the public CSS class names** (#5). Server-rendered markup needs them by definition, and it closes a request open since 2018 on its own.
4. **Then** server-render the thread and use htmx for votes, collapse, sorting and load-more, which is where it genuinely fits.

Steps 1 to 3 are worth doing whichever way step 4 goes, and they make step 4 a small decision rather than a rewrite undertaken on top of a compat layer.

**Comment by paskal** [comment]
Following up on the sequencing I proposed above, I worked through what server rendering would actually require against the current code. One finding changes the shape of the design, so it is worth recording here rather than leaving it to be rediscovered.

### The initial document render cannot be authenticated

Not "is hard to authenticate", and not a third-party cookie problem. It is closed by our own XSRF protection.

`go-pkgz/auth` rejects any token read from a cookie unless the request also carries an `X-XSRF-TOKEN` header matching the token's `jti` claim (`token/jwt.go:341-345` in v2.2.0). We set neither `DisableXSRF` nor `XSRFIgnoreMethods`, so the library default of an empty ignore list applies (`jwt.go:62`, applied at `:112-113`) and the method-based short circuit at `jwt.go:337` never fires for any method.

An iframe `src` navigation is a document request and cannot carry a custom header. The widget document therefore arrives anonymous with `AUTH_SAME_SITE=none`, with third-party cookies fully permitted, and on every browser. `authMiddleware.Trace` swallows the error (`middleware/auth.go:71-79`), so it degrades silently rather than failing.

There is one bypass, and it is worse than the constraint: `jwt.go:301` takes the token from a `?token=` query parameter, and because `fromCookie` stays false the XSRF check is skipped entirely. That would put a live JWT into the access log via `logInfoWithBody` (`rest.go:269`), into `Referer`, and into browser history. It is unreachable in any case, since the JWT cookie is `HttpOnly` (`jwt.go:280`) and no JavaScript can read it to build such a URL.

None of this blocks the direction. htmx fragments are XHR, and XHR does carry the header, exactly as `fetcher.ts` does today. What it means is that **anonymous-first is forced rather than chosen**: render the document anonymous, then hydrate over XHR.

### That makes hydration more expensive than it sounds

User identity threads through every node of the tree rather than sitting at its edges. `prepVotes` (`store/service/service.go:1096`) sets each comment's `Vote` from the requester and strips the voter map, `alterCommentCached` (`:1028`, `:1041`) blanks `User.IP` for non-admins, and the edit window, own-comment delete, admin controls and hidden users sit on top of that.

A signed-in reader therefore renders the tree twice: once anonymous in the document, once authenticated in the hydration fragment. The backend already shows the shape this takes, since `/find` keys its cache per user via `URLKeyWithUser` (`rest_public.go:101`, `rest.go:564`), with a separate `admin!!` key. An HTML cache fragments identically, so the CDN friendliness in the original post pays for logged-out readers and not for the rest.

### Other constraints worth knowing before anyone starts

- The dynamic route must sit outside `/api/v1`, not only outside `/web/`. `apiCSPMiddleware` (`rest.go:257`) replaces the CSP on everything mounted there with `default-src 'none'; sandbox; frame-ancestors 'none'` (`rest/image_headers.go:18`), which would sandbox the document and forbid framing it at all. The global `securityHeadersMiddleware` (`rest.go:221`) is what supplies `frame-ancestors` from `ALLOWED_HOSTS`.
- It must join the response cache's flush scopes. Comment and info responses are cached keyed by `Scopes(siteID, URL)` and flushed on writes (`rest_public.go:101`, `:165`); a route outside those scopes would serve stale comments after every post.
- Every private or admin fragment URL must carry `?site=`. `matchSiteID` (`middleware.go:196-202`) rejects an empty parameter with 403 rather than 401.
- The iframe has no intrinsic height. The parent sizes it from `postMessage({height})` (`utils/post-message.ts:84`), so every `hx-swap` needs a ResizeObserver feeding that message.
- The new route inherits the OAuth return stub, since `oauth.tsx:15` builds the return URL from the widget document's own path plus `?selfClose` and the iframe template closes the popup on arrival.
- Open routes are limited to 10 requests per second per IP (`rest.go:91`) against 20 for static `/web/` (`rest.go:518`), and a fragment UI multiplies request count per interaction against the tighter bucket.

On internationalisation, the objection that survived the 2020 discussion: `backend/app/templates` holds five templates today, four email plus an error page, and no catalogue machinery, so this is a from-zero build. The completeness check in `ci-frontend.yml` runs formatjs extraction over TypeScript sources and would be blind to strings living in Go templates, so replacement extraction tooling belongs to the same piece of work rather than to a follow-up.

### On the sequencing

Steps 1 to 3 from my previous comment stand and are worth doing whichever way step 4 goes. Step 1 is in review as #2175, which replaces `react-redux` with a small binding over Preact context and leaves `react-intl` as the only remaining consumer of the compat alias.

Two things have moved since that comment. `@remark42/api` was removed in #2172, so the point about a Svelte or Vue frontend already being unblocked now rests on the REST API and the documented embed contract rather than on a published package. And `pnpm audit` currently reports no known vulnerabilities, after the override floors were raised; the security cost of the toolchain is recurring maintenance rather than a standing exposure, which is a weaker form of that argument than the count I quoted.
